### PR TITLE
bugfix: buffer overflow exceptions because of 'z' compression

### PR DESCRIPTION
**Pull request by larsjaas**
include compression factor when allocating decoding buffer, to avoid buffer overflows

### CODE REVIEW

File: src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java [hunk 1]
```diff
@@ -105,4 +105,25 @@ public void overflowTest() {
         assertNotNull(Ascii85.decode(encodedString));
     }
 
+    @Test
```
Comment (fzakaria) [inline src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java:108]: please add a test that has 4+ bytes of zeros along with non-zero text?

File: src/main/java/com/github/fzakaria/ascii85/Ascii85.java [hunk 2]
```diff
@@ -91,10 +91,21 @@ public static String encode(byte[] payload) {
...
+        final int inputLength = chars.length();
+        int computedLength = 0;
+        for (int idx = 0; idx < inputLength; ++idx) {
+            if (chars.charAt(idx) != 'z') {
+                for (int c = 0; c < 5 && idx < inputLength; ++idx, ++c) {
+                    if (c > 0) ++computedLength;
+                }
+                --idx;
+            }
+            else computedLength += 4;
```
Comment (fzakaria) [inline src/main/java/com/github/fzakaria/ascii85/Ascii85.java:106]: please use braces

File: src/main/java/com/github/fzakaria/ascii85/Ascii85.java [hunk 3]
```diff
@@ -91,10 +91,21 @@ public static String encode(byte[] payload) {
...
             throw new IllegalArgumentException("You must provide a non-zero length input");
         }
-        //By using five ASCII characters to represent four bytes of binary data the encoded size ¹⁄₄ is larger than the original
-        BigDecimal decodedLength = BigDecimal.valueOf(chars.length()).multiply(BigDecimal.valueOf(4))
-                .divide(BigDecimal.valueOf(5));
-        ByteBuffer bytebuff = ByteBuffer.allocate(decodedLength.intValue());
+        // Because we perform compression when encoding four bytes of zeros to a single 'z', we need
+        // to scan through the input to compute the target length, instead of just subtracting 20% of
+        // the encoded text length.
+        final int inputLength = chars.length();
```
Comment (fzakaria) [inline src/main/java/com/github/fzakaria/ascii85/Ascii85.java:97]: Thanks for this bug find.

why not just do a string replace of `z` with `!!!!` and we keep the logic clearer?

File: src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java [hunk 4]
```diff
@@ -105,4 +105,25 @@ public void overflowTest() {
...
+        testZeroStringDecode(534);
+    }
+
+    public void testZeroStringDecode(int length) {
+        byte[] data = new byte[length];
+        for (int c = 0; c < length; c++) data[c] = 0;
+        try {
+            String encoded = Ascii85.encode(data);
+            byte[] decoded = Ascii85.decode(encoded);
+            assertEquals(length, decoded.length);
```
Comment (odisseus) [inline src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java:123]: Consider asserting equality of `data` and `decoded`.

File: src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java [hunk 5]
```diff
@@ -105,4 +105,25 @@ public void overflowTest() {
...
+        } catch (Exception e) {
```
Comment (odisseus) [inline src/test/java/com/github/fzakaria/ascii85/Ascii85Test.java:124]: This try-catch block is totally redundant. In fact, it replaces the explanation provided by JUnit with a generic "assertion failed" stack trace.

File: src/main/java/com/github/fzakaria/ascii85/Ascii85.java [hunk 6]
```diff
@@ -91,10 +91,21 @@ public static String encode(byte[] payload) {
...
+        int computedLength = 0;
```
Comment (odisseus) [inline src/main/java/com/github/fzakaria/ascii85/Ascii85.java:98]: Since the chunk into which each occurrence of 'z' is decoded is as long as a chunk for five ordinary characters,  the computation of decoded length can be rewritten like this:
```java
int zCount = 0;                                                
for(int i = 0 ; i < inputLength; i++) {                        
    if(chars.charAt(i) == 'z') ++ zCount;                      
}                                                              
int computedLength = 4 * zCount + 4 * (inputLength - zCount) / 5;
```

**Comment by odisseus** [comment]
This pull request would probably fix #4 and #7.

**Review by fzakaria — COMMENTED** [review]
Closing in favor of #10
